File: src/logics/PuyoType.ts
import { type ColoredPuyoAttr, PuyoAttr, isColoredPuyoAttr } from './PuyoAttr';

/** ぷよの型 */
export enum PuyoType {
  /** 赤ぷよ */
  Red = 1,
  /** 赤ぷよプラス */
  RedPlus = 2,
  /** 赤ぷよチャンス */
  RedChance = 3,
  /** 赤ぷよチャンスプラス */
  RedChancePlus = 4,
  /** 青ぷよ */
  Blue = 5,
  /** 青ぷよプラス */
  BluePlus = 6,
  /** 青ぷよチャンス */
  BlueChance = 7,
  /** 青ぷよチャンスプラス */
  BlueChancePlus = 8,
  /** 緑ぷよ */
  Green = 9,
  /** 緑ぷよプラス */
  GreenPlus = 10,
  /** 緑ぷよチャンス */
  GreenChance = 11,
  /** 緑ぷよチャンスプラス */
  GreenChancePlus = 12,
  /** 黄ぷよ */
  Yellow = 13,
  /** 黄ぷよプラス */
  YellowPlus = 14,
  /** 黄ぷよチャンス */
  YellowChance = 15,
  /** 黄ぷよチャンスプラス */
  YellowChancePlus = 16,
  /** 紫ぷよ */
  Purple = 17,
  /** 紫ぷよプラス */
  PurplePlus = 18,
  /** 紫ぷよチャンス */
  PurpleChance = 19,
  /** 紫ぷよチャンスプラス */
  PurpleChancePlus = 20,
  /** ハートぷよ */
  Heart = 21,
  /** プリズムぷよ */
  Prism = 22,
  /** おじゃまぷよ */
  Ojama = 23,
  /** 固ぷよ */
  Kata = 24,
  /** パディングぷよ (連鎖をとめるための消せないぷよ) */
  Padding = 25
}

export const puyoTypeMap: ReadonlyMap<PuyoType, string> = new Map<
  PuyoType,
  string
>([
  [PuyoType.Red, '赤'],
  [PuyoType.RedPlus, '赤+'],
  [PuyoType.RedChance, '赤ch'],
  [PuyoType.RedChancePlus, '赤ch+'],
  [PuyoType.Blue, '青'],
  [PuyoType.BluePlus, '青+'],
  [PuyoType.BlueChance, '青ch'],
  [PuyoType.BlueChancePlus, '青ch+'],
  [PuyoType.Green, '緑'],
  [PuyoType.GreenPlus, '緑+'],
  [PuyoType.GreenChance, '緑ch'],
  [PuyoType.GreenChancePlus, '緑ch+'],
  [PuyoType.Yellow, '黄'],
  [PuyoType.YellowPlus, '黄+'],
  [PuyoType.YellowChance, '黄ch'],
  [PuyoType.YellowChancePlus, '黄ch+'],
  [PuyoType.Purple, '紫'],
  [PuyoType.PurplePlus, '紫+'],
  [PuyoType.PurpleChance, '紫ch'],
  [PuyoType.PurpleChancePlus, '紫ch+'],
  [PuyoType.Heart, 'ハート'],
  [PuyoType.Prism, 'プリズム'],
  [PuyoType.Ojama, 'おじゃま'],
  [PuyoType.Kata, '固'],
  [PuyoType.Padding, 'パティング']
]);

export const getPuyoTypeName = (puyoType: PuyoType | undefined): string => {
  return puyoTypeMap.get(puyoType!) || '';
};

/** 色ぷよの型 (プリズムは含まない) */
export type ColoredPuyoType =
  | PuyoType.Red
  | PuyoType.RedPlus
  | PuyoType.RedChance
  | PuyoType.RedChancePlus
  | PuyoType.Blue
  | PuyoType.BluePlus
  | PuyoType.BlueChance
  | PuyoType.BlueChancePlus
  | PuyoType.Green
  | PuyoType.GreenPlus
  | PuyoType.GreenChance
  | PuyoType.GreenChancePlus
  | PuyoType.Yellow
  | PuyoType.YellowPlus
  | PuyoType.YellowChance
  | PuyoType.YellowChancePlus
  | PuyoType.Purple
  | PuyoType.PurplePlus
  | PuyoType.PurpleChance
  | PuyoType.PurpleChancePlus;

/**
 * 色ぷよかどうか (プリズムは含まない)
 * @param puyoType ぷよの型
 * @returns
 */
export const isColoredPuyoType = (
  puyoType: PuyoType
): puyoType is ColoredPuyoType => {
  return puyoType >= PuyoType.Red && puyoType <= PuyoType.PurpleChancePlus;
};

/**
 * プラスぷよかどうか
 * @param puyoType ぷよの型
 * @returns
 */
export const isPlusPuyo = (
  puyoType: PuyoType | undefined
): puyoType is
  | PuyoType.RedPlus
  | PuyoType.RedChancePlus
  | PuyoType.BluePlus
  | PuyoType.BlueChancePlus
  | PuyoType.GreenPlus
  | PuyoType.GreenChancePlus
  | PuyoType.YellowPlus
  | PuyoType.YellowChancePlus
  | PuyoType.PurplePlus
  | PuyoType.PurpleChancePlus => {
  switch (puyoType) {
    case PuyoType.RedPlus:
    case PuyoType.RedChancePlus:
    case PuyoType.BluePlus:
    case PuyoType.BlueChancePlus:
    case PuyoType.GreenPlus:
    case PuyoType.GreenChancePlus:
    case PuyoType.YellowPlus:
    case PuyoType.YellowChancePlus:
    case PuyoType.PurplePlus:
    case PuyoType.PurpleChancePlus:
      return true;
  }
  return false;
};

/**
 * チャンスぷよかどうか
 * @param puyoType ぷよの型
 * @returns
 */
export const isChancePuyo = (
  puyoType: PuyoType
): puyoType is
  | PuyoType.RedChance
  | PuyoType.RedChancePlus
  | PuyoType.BlueChance
  | PuyoType.BlueChancePlus
  | PuyoType.GreenChance
  | PuyoType.GreenChancePlus
  | PuyoType.YellowChance
  | PuyoType.YellowChancePlus
  | PuyoType.PurpleChance
  | PuyoType.PurpleChancePlus => {
  switch (puyoType) {
    case PuyoType.RedChance:
    case PuyoType.RedChancePlus:
    case PuyoType.BlueChance:
    case PuyoType.BlueChancePlus:
    case PuyoType.GreenChance:
    case PuyoType.GreenChancePlus:
    case PuyoType.YellowChance:
    case PuyoType.YellowChancePlus:
    case PuyoType.PurpleChance:
    case PuyoType.PurpleChancePlus:
      return true;
  }
  return false;
};

/** なぞり可能なぷよかどうか (星魔のような塗りは考慮していない) */
export const isTraceablePuyo = (
  puyoType: PuyoType | undefined
): puyoType is ColoredPuyoType | PuyoType.Heart | PuyoType.Prism => {
  if (!puyoType) {
    return false;
  }

  switch (puyoType) {
    case PuyoType.Ojama:
    case PuyoType.Kata:
    case PuyoType.Padding:
      return false;
    default:
      return true;
  }
};

/** ぷよの属性 */
export const getPuyoAttr = (
  puyoType: PuyoType | undefined
): PuyoAttr | undefined => {
  switch (puyoType) {
    case PuyoType.Red:
    case PuyoType.RedPlus:
    case PuyoType.RedChance:
    case PuyoType.RedChancePlus:
      return PuyoAttr.Red;

    case PuyoType.Blue:
    case PuyoType.BluePlus:
    case PuyoType.BlueChance:
    case PuyoType.BlueChancePlus:
      return PuyoAttr.Blue;

    case PuyoType.Green:
    case PuyoType.GreenPlus:
    case PuyoType.GreenChance:
    case PuyoType.GreenChancePlus:
      return PuyoAttr.Green;

    case PuyoType.Yellow:
    case PuyoType.YellowPlus:
    case PuyoType.YellowChance:
    case PuyoType.YellowChancePlus:
      return PuyoAttr.Yellow;

    case PuyoType.Purple:
    case PuyoType.PurplePlus:
    case PuyoType.PurpleChance:
    case PuyoType.PurpleChancePlus:
      return PuyoAttr.Purple;

    case PuyoType.Heart:
      return PuyoAttr.Heart;

    case PuyoType.Prism:
      return PuyoAttr.Prism;

    case PuyoType.Ojama:
      return PuyoAttr.Ojama;

    case PuyoType.Kata:
      return PuyoAttr.Kata;

    case PuyoType.Padding:
      return PuyoAttr.Padding;
  }
};

/**
 * 色ぷよであればプラスもチャンスも付いていない通常の色ぷよに変換する。
 * @param puyoType
 * @returns
 */
export const toNormalColoredType = (puyoType: PuyoType) => {
  if (!isColoredPuyoType(puyoType)) {
    return puyoType;
  }
  const attr = getPuyoAttr(puyoType) as ColoredPuyoAttr;

  switch (attr) {
    case PuyoAttr.Red:
      return PuyoType.Red;
    case PuyoAttr.Blue:
      return PuyoType.Blue;
    case PuyoAttr.Green:
      return PuyoType.Green;
    case PuyoAttr.Yellow:
      return PuyoType.Yellow;
    case PuyoAttr.Purple:
      return PuyoType.Purple;
  }
};

/**
 * 色ぷよであればチャンス付与する。
 * @param puyoType
 * @returns
 */
export const toChanceColoredType = (puyoType: PuyoType) => {
  if (!isColoredPuyoType(puyoType)) {
    return puyoType;
  }

  const isPlus = isPlusPuyo(puyoType);
  const attr = getPuyoAttr(puyoType) as ColoredPuyoAttr;

  switch (attr) {
    case PuyoAttr.Red:
      return isPlus ? PuyoType.RedChancePlus : PuyoType.RedChance;
    case PuyoAttr.Blue:
      return isPlus ? PuyoType.BlueChancePlus : PuyoType.BlueChance;
    case PuyoAttr.Green:
      return isPlus ? PuyoType.GreenChancePlus : PuyoType.GreenChance;
    case PuyoAttr.Yellow:
      return isPlus ? PuyoType.YellowChancePlus : PuyoType.YellowChance;
    case PuyoAttr.Purple:
      return isPlus ? PuyoType.PurpleChancePlus : PuyoType.PurpleChance;
  }
};

/**
 * 色ぷよであればプラスを付与する。
 * @param puyoType
 * @returns
 */
export const toPlusColoredType = (puyoType: PuyoType) => {
  if (!isColoredPuyoType(puyoType)) {
    return puyoType;
  }

  const isChance = isChancePuyo(puyoType);
  const attr = getPuyoAttr(puyoType) as ColoredPuyoAttr;

  switch (attr) {
    case PuyoAttr.Red:
      return isChance ? PuyoType.RedChancePlus : PuyoType.RedPlus;
    case PuyoAttr.Blue:
      return isChance ? PuyoType.BlueChancePlus : PuyoType.BluePlus;
    case PuyoAttr.Green:
      return isChance ? PuyoType.GreenChancePlus : PuyoType.GreenPlus;
    case PuyoAttr.Yellow:
      return isChance ? PuyoType.YellowChancePlus : PuyoType.YellowPlus;
    case PuyoAttr.Purple:
      return isChance ? PuyoType.PurpleChancePlus : PuyoType.PurplePlus;
  }
};

/**
 * ぷよの型を指定の属性に変換する。
 * @param puyoType
 * @param toAttr
 * @returns
 */
export const convertPuyoType = (
  puyoType: PuyoType,
  toAttr: PuyoAttr
): PuyoType => {
  if (puyoType === PuyoType.Padding) {
    return puyoType;
  }

  if (isColoredPuyoType(puyoType)) {
    if (isColoredPuyoAttr(toAttr)) {
      const isPlusTerm = isPlusPuyo(puyoType) ? 1 : 0;
      const isChanceTerm = isChancePuyo(puyoType) ? 2 : 0;

      let result!: PuyoType;

      switch (toAttr) {
        case PuyoAttr.Red:
          result = PuyoType.Red;
          break;
        case PuyoAttr.Blue:
          result = PuyoType.Blue;
          break;
        case PuyoAttr.Green:
          result = PuyoType.Green;
          break;
        case PuyoAttr.Yellow:
          result = PuyoType.Yellow;
          break;
        case PuyoAttr.Purple:
          result = PuyoType.Purple;
      }

      result += isPlusTerm + isChanceTerm;

      return result;
    }

    switch (toAttr) {
      case PuyoAttr.Heart:
        return PuyoType.Heart;
      case PuyoAttr.Prism:
        return PuyoType.Prism;
      case PuyoAttr.Ojama:
        return PuyoType.Ojama;
      case PuyoAttr.Kata:
        return PuyoType.Kata;
    }
  }

  switch (toAttr) {
    case PuyoAttr.Red:
      return PuyoType.Red;
    case PuyoAttr.Blue:
      return PuyoType.Blue;
    case PuyoAttr.Green:
      return PuyoType.Green;
    case PuyoAttr.Yellow:
      return PuyoType.Yellow;
    case PuyoAttr.Purple:
      return PuyoType.Purple;
    case PuyoAttr.Heart:
      return PuyoType.Heart;
    case PuyoAttr.Prism:
      return PuyoType.Prism;
    case PuyoAttr.Ojama:
      return PuyoType.Ojama;
    case PuyoAttr.Kata:
      return PuyoType.Kata;
    case PuyoAttr.Padding:
      return PuyoType.Padding;
  }
};

/** TODO: SVG化したら削除 */
export const getPuyoRgb = (type: PuyoType) => {
  const attr = getPuyoAttr(type);

  switch (attr) {
    case PuyoAttr.Red:
      return '#c00';
    case PuyoAttr.Blue:
      return '#00c';
    case PuyoAttr.Green:
      return '#0c0';
    case PuyoAttr.Yellow:
      return 'yellow';
    case PuyoAttr.Purple:
      return '#808';
    case PuyoAttr.Heart:
      return 'pink';
    case PuyoAttr.Prism:
      return '#fff';
    case PuyoAttr.Ojama:
      return '#666';
    case PuyoAttr.Kata:
      return '#333';
    case PuyoAttr.Padding:
      return '#000';
    default:
      throw new Error('The attr of puyo is unknown.');
  }
};

File: packages/solver-wasm/src/puyo_type.rs
use num_derive::{FromPrimitive, ToPrimitive};
use num_traits::{FromPrimitive, ToPrimitive};
use serde_repr::{Deserialize_repr, Serialize_repr};

use crate::puyo_attr::*;

#[derive(
    Debug, Copy, Clone, PartialEq, Eq, FromPrimitive, ToPrimitive, Serialize_repr, Deserialize_repr,
)]
#[repr(u8)]
pub enum PuyoType {
    Red = 1,
    RedPlus = 2,
    RedChance = 3,
    RedChancePlus = 4,
    Blue = 5,
    BluePlus = 6,
    BlueChance = 7,
    BlueChancePlus = 8,
    Green = 9,
    GreenPlus = 10,
    GreenChance = 11,
    GreenChancePlus = 12,
    Yellow = 13,
    YellowPlus = 14,
    YellowChance = 15,
    YellowChancePlus = 16,
    Purple = 17,
    PurplePlus = 18,
    PurpleChance = 19,
    PurpleChancePlus = 20,
    Heart = 21,
    Prism = 22,
    Ojama = 23,
    Kata = 24,
    Padding = 25,
}

pub fn is_colored_type(puyo_type: PuyoType) -> bool {
    match puyo_type {
        PuyoType::Red => true,
        PuyoType::RedPlus => true,
        PuyoType::RedChance => true,
        PuyoType::RedChancePlus => true,
        PuyoType::Blue => true,
        PuyoType::BluePlus => true,
        PuyoType::BlueChance => true,
        PuyoType::BlueChancePlus => true,
        PuyoType::Green => true,
        PuyoType::GreenPlus => true,
        PuyoType::GreenChance => true,
        PuyoType::GreenChancePlus => true,
        PuyoType::Yellow => true,
        PuyoType::YellowPlus => true,
        PuyoType::YellowChance => true,
        PuyoType::YellowChancePlus => true,
        PuyoType::Purple => true,
        PuyoType::PurplePlus => true,
        PuyoType::PurpleChance => true,
        PuyoType::PurpleChancePlus => true,
        _ => false,
    }
}

pub fn is_plus_type(puyo_type: PuyoType) -> bool {
    match puyo_type {
        PuyoType::RedPlus => true,
        PuyoType::RedChancePlus => true,
        PuyoType::BluePlus => true,
        PuyoType::BlueChancePlus => true,
        PuyoType::GreenPlus => true,
        PuyoType::GreenChancePlus => true,
        PuyoType::YellowPlus => true,
        PuyoType::YellowChancePlus => true,
        PuyoType::PurplePlus => true,
        PuyoType::PurpleChancePlus => true,
        _ => false,
    }
}

pub fn is_chance_type(puyo_type: PuyoType) -> bool {
    match puyo_type {
        PuyoType::RedChance => true,
        PuyoType::RedChancePlus => true,
        PuyoType::BlueChance => true,
        PuyoType::BlueChancePlus => true,
        PuyoType::GreenChance => true,
        PuyoType::GreenChancePlus => true,
        PuyoType::YellowChance => true,
        PuyoType::YellowChancePlus => true,
        PuyoType::PurpleChance => true,
        PuyoType::PurpleChancePlus => true,
        _ => false,
    }
}

pub fn is_traceable_type(puyo_type: PuyoType) -> bool {
    match puyo_type {
        PuyoType::Ojama => false,
        PuyoType::Kata => false,
        PuyoType::Padding => false,
        _ => true,
    }
}

pub fn get_attr(puyo_type: PuyoType) -> PuyoAttr {
    match puyo_type {
        PuyoType::Red => PuyoAttr::Red,
        PuyoType::RedPlus => PuyoAttr::Red,
        PuyoType::RedChance => PuyoAttr::Red,
        PuyoType::RedChancePlus => PuyoAttr::Red,
        PuyoType::Blue => PuyoAttr::Blue,
        PuyoType::BluePlus => PuyoAttr::Blue,
        PuyoType::BlueChance => PuyoAttr::Blue,
        PuyoType::BlueChancePlus => PuyoAttr::Blue,
        PuyoType::Green => PuyoAttr::Green,
        PuyoType::GreenPlus => PuyoAttr::Green,
        PuyoType::GreenChance => PuyoAttr::Green,
        PuyoType::GreenChancePlus => PuyoAttr::Green,
        PuyoType::Yellow => PuyoAttr::Yellow,
        PuyoType::YellowPlus => PuyoAttr::Yellow,
        PuyoType::YellowChance => PuyoAttr::Yellow,
        PuyoType::YellowChancePlus => PuyoAttr::Yellow,
        PuyoType::Purple => PuyoAttr::Purple,
        PuyoType::PurplePlus => PuyoAttr::Purple,
        PuyoType::PurpleChance => PuyoAttr::Purple,
        PuyoType::PurpleChancePlus => PuyoAttr::Purple,
        PuyoType::Heart => PuyoAttr::Heart,
        PuyoType::Prism => PuyoAttr::Prism,
        PuyoType::Ojama => PuyoAttr::Ojama,
        PuyoType::Kata => PuyoAttr::Kata,
        PuyoType::Padding => PuyoAttr::Padding,
    }
}

pub fn convert_type(puyo_type: PuyoType, attr: PuyoAttr) -> PuyoType {
    if is_colored_attr(attr) {
        let plus_term = if is_plus_type(puyo_type) { 1 } else { 0 };
        let chance_term = if is_chance_type(puyo_type) { 2 } else { 0 };
        let enhance = plus_term + chance_term;

        return match attr {
            PuyoAttr::Red => PuyoType::from_i32(PuyoType::Red.to_i32().unwrap() + enhance).unwrap(),
            PuyoAttr::Blue => {
                PuyoType::from_i32(PuyoType::Blue.to_i32().unwrap() + enhance).unwrap()
            }
            PuyoAttr::Green => {
                PuyoType::from_i32(PuyoType::Green.to_i32().unwrap() + enhance).unwrap()
            }
            PuyoAttr::Yellow => {
                PuyoType::from_i32(PuyoType::Yellow.to_i32().unwrap() + enhance).unwrap()
            }
            PuyoAttr::Purple => {
                PuyoType::from_i32(PuyoType::Purple.to_i32().unwrap() + enhance).unwrap()
            }
            _ => PuyoType::Padding,
        };
    }
    return match attr {
        PuyoAttr::Heart => PuyoType::Heart,
        PuyoAttr::Prism => PuyoType::Prism,
        PuyoAttr::Ojama => PuyoType::Ojama,
        PuyoAttr::Kata => PuyoType::Kata,
        PuyoAttr::Padding => PuyoType::Padding,
        _ => PuyoType::Padding,
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_colored_type() {
        assert_eq!(is_colored_type(PuyoType::Red), true);
        assert_eq!(is_colored_type(PuyoType::RedPlus), true);
        assert_eq!(is_colored_type(PuyoType::RedChance), true);
        assert_eq!(is_colored_type(PuyoType::RedChancePlus), true);
        assert_eq!(is_colored_type(PuyoType::Blue), true);
        assert_eq!(is_colored_type(PuyoType::BluePlus), true);
        assert_eq!(is_colored_type(PuyoType::BlueChance), true);
        assert_eq!(is_colored_type(PuyoType::BlueChancePlus), true);
        assert_eq!(is_colored_type(PuyoType::Green), true);
        assert_eq!(is_colored_type(PuyoType::GreenPlus), true);
        assert_eq!(is_colored_type(PuyoType::GreenChance), true);
        assert_eq!(is_colored_type(PuyoType::GreenChancePlus), true);
        assert_eq!(is_colored_type(PuyoType::Yellow), true);
        assert_eq!(is_colored_type(PuyoType::YellowPlus), true);
        assert_eq!(is_colored_type(PuyoType::YellowChance), true);
        assert_eq!(is_colored_type(PuyoType::YellowChancePlus), true);
        assert_eq!(is_colored_type(PuyoType::Purple), true);
        assert_eq!(is_colored_type(PuyoType::PurplePlus), true);
        assert_eq!(is_colored_type(PuyoType::PurpleChance), true);
        assert_eq!(is_colored_type(PuyoType::PurpleChancePlus), true);
        assert_eq!(is_colored_type(PuyoType::Heart), false);
        assert_eq!(is_colored_type(PuyoType::Prism), false);
        assert_eq!(is_colored_type(PuyoType::Ojama), false);
        assert_eq!(is_colored_type(PuyoType::Kata), false);
        assert_eq!(is_colored_type(PuyoType::Padding), false);
    }

    #[test]
    fn test_is_plus_type() {
        assert_eq!(is_plus_type(PuyoType::Red), false);
        assert_eq!(is_plus_type(PuyoType::RedPlus), true);
        assert_eq!(is_plus_type(PuyoType::RedChance), false);
        assert_eq!(is_plus_type(PuyoType::RedChancePlus), true);
        assert_eq!(is_plus_type(PuyoType::Blue), false);
        assert_eq!(is_plus_type(PuyoType::BluePlus), true);
        assert_eq!(is_plus_type(PuyoType::BlueChance), false);
        assert_eq!(is_plus_type(PuyoType::BlueChancePlus), true);
        assert_eq!(is_plus_type(PuyoType::Green), false);
        assert_eq!(is_plus_type(PuyoType::GreenPlus), true);
        assert_eq!(is_plus_type(PuyoType::GreenChance), false);
        assert_eq!(is_plus_type(PuyoType::GreenChancePlus), true);
        assert_eq!(is_plus_type(PuyoType::Yellow), false);
        assert_eq!(is_plus_type(PuyoType::YellowPlus), true);
        assert_eq!(is_plus_type(PuyoType::YellowChance), false);
        assert_eq!(is_plus_type(PuyoType::YellowChancePlus), true);
        assert_eq!(is_plus_type(PuyoType::Purple), false);
        assert_eq!(is_plus_type(PuyoType::PurplePlus), true);
        assert_eq!(is_plus_type(PuyoType::PurpleChance), false);
        assert_eq!(is_plus_type(PuyoType::PurpleChancePlus), true);
        assert_eq!(is_plus_type(PuyoType::Heart), false);
        assert_eq!(is_plus_type(PuyoType::Prism), false);
        assert_eq!(is_plus_type(PuyoType::Ojama), false);
        assert_eq!(is_plus_type(PuyoType::Kata), false);
        assert_eq!(is_plus_type(PuyoType::Padding), false);
    }

    #[test]
    fn test_is_chance_type() {
        assert_eq!(is_chance_type(PuyoType::Red), false);
        assert_eq!(is_chance_type(PuyoType::RedPlus), false);
        assert_eq!(is_chance_type(PuyoType::RedChance), true);
        assert_eq!(is_chance_type(PuyoType::RedChancePlus), true);
        assert_eq!(is_chance_type(PuyoType::Blue), false);
        assert_eq!(is_chance_type(PuyoType::BluePlus), false);
        assert_eq!(is_chance_type(PuyoType::BlueChance), true);
        assert_eq!(is_chance_type(PuyoType::BlueChancePlus), true);
        assert_eq!(is_chance_type(PuyoType::Green), false);
        assert_eq!(is_chance_type(PuyoType::GreenPlus), false);
        assert_eq!(is_chance_type(PuyoType::GreenChance), true);
        assert_eq!(is_chance_type(PuyoType::GreenChancePlus), true);
        assert_eq!(is_chance_type(PuyoType::Yellow), false);
        assert_eq!(is_chance_type(PuyoType::YellowPlus), false);
        assert_eq!(is_chance_type(PuyoType::YellowChance), true);
        assert_eq!(is_chance_type(PuyoType::YellowChancePlus), true);
        assert_eq!(is_chance_type(PuyoType::Purple), false);
        assert_eq!(is_chance_type(PuyoType::PurplePlus), false);
        assert_eq!(is_chance_type(PuyoType::PurpleChance), true);
        assert_eq!(is_chance_type(PuyoType::PurpleChancePlus), true);
        assert_eq!(is_chance_type(PuyoType::Heart), false);
        assert_eq!(is_chance_type(PuyoType::Prism), false);
        assert_eq!(is_chance_type(PuyoType::Ojama), false);
        assert_eq!(is_chance_type(PuyoType::Kata), false);
        assert_eq!(is_chance_type(PuyoType::Padding), false);
    }

    #[test]
    fn test_is_traceable_type() {
        assert_eq!(is_traceable_type(PuyoType::Red), true);
        assert_eq!(is_traceable_type(PuyoType::RedPlus), true);
        assert_eq!(is_traceable_type(PuyoType::RedChance), true);
        assert_eq!(is_traceable_type(PuyoType::RedChancePlus), true);
        assert_eq!(is_traceable_type(PuyoType::Blue), true);
        assert_eq!(is_traceable_type(PuyoType::BluePlus), true);
        assert_eq!(is_traceable_type(PuyoType::BlueChance), true);
        assert_eq!(is_traceable_type(PuyoType::BlueChancePlus), true);
        assert_eq!(is_traceable_type(PuyoType::Green), true);
        assert_eq!(is_traceable_type(PuyoType::GreenPlus), true);
        assert_eq!(is_traceable_type(PuyoType::GreenChance), true);
        assert_eq!(is_traceable_type(PuyoType::GreenChancePlus), true);
        assert_eq!(is_traceable_type(PuyoType::Yellow), true);
        assert_eq!(is_traceable_type(PuyoType::YellowPlus), true);
        assert_eq!(is_traceable_type(PuyoType::YellowChance), true);
        assert_eq!(is_traceable_type(PuyoType::YellowChancePlus), true);
        assert_eq!(is_traceable_type(PuyoType::Purple), true);
        assert_eq!(is_traceable_type(PuyoType::PurplePlus), true);
        assert_eq!(is_traceable_type(PuyoType::PurpleChance), true);
        assert_eq!(is_traceable_type(PuyoType::PurpleChancePlus), true);
        assert_eq!(is_traceable_type(PuyoType::Heart), true);
        assert_eq!(is_traceable_type(PuyoType::Prism), true);
        assert_eq!(is_traceable_type(PuyoType::Ojama), false);
        assert_eq!(is_traceable_type(PuyoType::Kata), false);
        assert_eq!(is_traceable_type(PuyoType::Padding), false);
    }

    #[test]
    fn test_get_attr() {
        assert_eq!(get_attr(PuyoType::Red), PuyoAttr::Red);
        assert_eq!(get_attr(PuyoType::RedPlus), PuyoAttr::Red);
        assert_eq!(get_attr(PuyoType::RedChance), PuyoAttr::Red);
        assert_eq!(get_attr(PuyoType::RedChancePlus), PuyoAttr::Red);
        assert_eq!(get_attr(PuyoType::Blue), PuyoAttr::Blue);
        assert_eq!(get_attr(PuyoType::BluePlus), PuyoAttr::Blue);
        assert_eq!(get_attr(PuyoType::BlueChance), PuyoAttr::Blue);
        assert_eq!(get_attr(PuyoType::BlueChancePlus), PuyoAttr::Blue);
        assert_eq!(get_attr(PuyoType::Green), PuyoAttr::Green);
        assert_eq!(get_attr(PuyoType::GreenPlus), PuyoAttr::Green);
        assert_eq!(get_attr(PuyoType::GreenChance), PuyoAttr::Green);
        assert_eq!(get_attr(PuyoType::GreenChancePlus), PuyoAttr::Green);
        assert_eq!(get_attr(PuyoType::Yellow), PuyoAttr::Yellow);
        assert_eq!(get_attr(PuyoType::YellowPlus), PuyoAttr::Yellow);
        assert_eq!(get_attr(PuyoType::YellowChance), PuyoAttr::Yellow);
        assert_eq!(get_attr(PuyoType::YellowChancePlus), PuyoAttr::Yellow);
        assert_eq!(get_attr(PuyoType::Purple), PuyoAttr::Purple);
        assert_eq!(get_attr(PuyoType::PurplePlus), PuyoAttr::Purple);
        assert_eq!(get_attr(PuyoType::PurpleChance), PuyoAttr::Purple);
        assert_eq!(get_attr(PuyoType::PurpleChancePlus), PuyoAttr::Purple);
        assert_eq!(get_attr(PuyoType::Heart), PuyoAttr::Heart);
        assert_eq!(get_attr(PuyoType::Prism), PuyoAttr::Prism);
        assert_eq!(get_attr(PuyoType::Ojama), PuyoAttr::Ojama);
        assert_eq!(get_attr(PuyoType::Kata), PuyoAttr::Kata);
        assert_eq!(get_attr(PuyoType::Padding), PuyoAttr::Padding);
    }

    #[test]
    fn test_convert_type() {
        assert_eq!(
            convert_type(PuyoType::RedChancePlus, PuyoAttr::Blue),
            PuyoType::BlueChancePlus
        );
        assert_eq!(
            convert_type(PuyoType::Blue, PuyoAttr::Green),
            PuyoType::Green
        );
        assert_eq!(
            convert_type(PuyoType::YellowPlus, PuyoAttr::Purple),
            PuyoType::PurplePlus
        );
        assert_eq!(
            convert_type(PuyoType::PurpleChance, PuyoAttr::Red),
            PuyoType::RedChance
        );
        assert_eq!(
            convert_type(PuyoType::RedChancePlus, PuyoAttr::Heart),
            PuyoType::Heart
        );
        assert_eq!(
            convert_type(PuyoType::Heart, PuyoAttr::Prism),
            PuyoType::Prism
        );
    }
}
